idb/migration/migration_test.go

```diff
@@ -329,3 +330,114 @@ func TestSuccessfulMigration(t *testing.T) {
 		})
 	}
 }
+
+// TestAvailabilityChannelCloses tests that the migration object closes the availability
+// channel when blocking migrations finish.
+func TestAvailabilityChannelCloses(t *testing.T) {
+	// Migration 2 reads on this channel.
+	migrationTwoChannel := make(chan struct{})
+	defer func() {
+		migrationTwoChannel <- struct{}{}
+	}()
+
+	tasks := []Task{
+		{
+			MigrationID: 1,
+			Handler: func() error {
+				return nil
+			},
+			DBUnavailable: true,
+		},
+		{
+			MigrationID: 2,
+			Handler: func() error {
+				<-migrationTwoChannel
+				return nil
+			},
+		},
+	}
+
+	m, err := MakeMigration(tasks, nil)
+	require.NoError(t, err)
+
+	availableCh := m.RunMigrations()
+	select {
+	case _, ok := <-availableCh:
+		assert.False(t, ok)
+	case <-time.After(10 * time.Millisecond):
+		assert.Fail(t, "channel must be closed")
+	}
+}
+
+// TestAvailabilityChannelClosesNoMigrations tests that the migration object closes
+// the availability channel when last migration, which is blocking, finishes.
+// If it is not the case, this test will deadlock (and timeout).
+func TestAvailabilityChannelClosesNoMigrations(t *testing.T) {
+	tasks := []Task{
+		{
+			MigrationID: 1,
+			Handler: func() error {
+				return nil
+			},
+			DBUnavailable: true,
+		},
+	}
+
+	m, err := MakeMigration(tasks, nil)
+	require.NoError(t, err)
+
+	availableCh := m.RunMigrations()
+	select {
+	case _, ok := <-availableCh:
+		assert.False(t, ok)
+	case <-time.After(10 * time.Millisecond):
+		assert.Fail(t, "channel must be closed")
+	}
+}
+
+// TestAvailabilityChannelClosesBlockingMigrationLast tests that the migration object
+// closes when there are no migrations. If it is not the case, this test will
+// deadlock (and timeout).
+func TestAvailabilityChannelClosesBlockingMigrationLast(t *testing.T) {
+	m, err := MakeMigration([]Task{}, nil)
+	require.NoError(t, err)
+
+	availableCh := m.RunMigrations()
+	select {
+	case _, ok := <-availableCh:
+		assert.False(t, ok)
+	case <-time.After(10 * time.Millisecond):
+		assert.Fail(t, "channel must be closed")
+	}
+}
+
+// TestAvailabilityChannelDoesNotCloseEarly tests that the migration object closes the availability
+// channel only after all blocking migrations are run.
+func TestAvailabilityChannelDoesNotCloseEarly(t *testing.T) {
```

Comment:
No, it's different. That one tests that the channel is closed not too late. This one tests that the channel is closed not too early.

Reply:
I see. In that case, I think that you could use a channel like the first test instead of sleeping in the handler.

Reply:
I'm not sure how a channel can be used here. If you have an idea, I can make the change in another PR.

Reply:
This is splitting hairs, but I believe it would be the same as `TestAvailabilityChannelCloses`. Instead of having the 10ms sleep it would be stopped by the channel until after the test finishes and the migration is released by the deferred channel write..

Reply:
(Pseudo)-code would be helpful for me to understand.

Reply:
```diff
	// Migration 2 reads on this channel.
+	migrationTwoChannel := make(chan struct{})
+	defer func() {
+		migrationTwoChannel <- struct{}{}
+	}()

	tasks := []Task{
		{
			MigrationID: 1,
			Handler: func() error {
				return nil
			},
		},
		{
			MigrationID: 2,
			Handler: func() error {
+				<-migrationTwoChannel
-				time.Sleep(10 * time.Millisecond)
				return nil
			},
			DBUnavailable: true,
		},
	}

	m, err := MakeMigration(tasks, nil)
	require.NoError(t, err)

	availableCh := m.RunMigrations()
	select {
	case <-availableCh:
		assert.Fail(t, "availability channel closed before migrations finish running")
	case <-time.After(5 * time.Millisecond):
	}

Reply:
Thanks. Agree. https://github.com/algorand/indexer/pull/592